publications/radstack/radstack.tex

```diff
@@ -0,0 +1,1148 @@
+availability, processed events from Samza are transmitted concurrently to two
+real-time nodes. Both nodes receive the same copy of data, and effectively act
+as replicas of each other. The Druid broker does not distinguish between the
+two copies of data, and queries will go to either node. When handoff occurs,
+both real-time nodes race to hand off the segments they’ve created. The segment
+that is pushed into deep storage first will be the one that is used for
+historical querying, and once that segment is loaded on the historical nodes,
+both real-time nodes will drop their versions of the same segment.
+
+\subsection{Batch Processing}
+Our batch processing pipeline is composed a multi-stage
+MapReduce\cite{dean2008mapreduce} pipeline. The first set of jobs mirrors our
+streaming processing pipeline in that it transforms data and joins relevant
+events in preparation for the is data to be loaded into Druid. The second set
+of jobs is responsible for directly creating immutable Druid segments. A final
+set of MapReduce jobs that reads our processed data from the first set and
+indexes this data to create Druid segments. The indexing code for both
+streaming and batch ingestion in Druid is shared between the two modes of
+ingestion. These segments are then uploaded to deep storage and registered with
+the  metadata store. Druid will proceed to load the batch generated segments.
+
+The batch process typically runs much less frequently than the real-time
+process, and may run many hours or even days after raw events have been
+delivered. The wait is necessary for severely delayed events, and to ensure
+that the raw data is indeed complete. 
+
+Segments generated by the batch process are versioned by the start time of the
+process. Hence, segments created by batch processing will have a version
+identifier that is greater than segments created by real-time processing. When
+these batch created segments are loaded in the cluster, they atomically replace
+segments created by real-time processing for their processed interval. Hence,
+soon after batch processing completes, Druid queries begin reflecting
+batch-originated data rather than real-time-originated data.
+
+
+Our ingestion model is hopefully more clear at this point. We use the streaming
+data pipeline described in Section\ref{sec:streamprocessing} to deliver
+immediate insights on events as they are occurring, and the batch data pipeline
+described in this section to provide the accurate copy of the data. The batch
+process typically runs much less frequently than the real-time process, and may
+run many hours or even days after raw events have been delivered. The wait is
+necessary for severely delayed events, and to ensure that the raw data is
+indeed complete. 
+
+\subsection{Starfire}
+To avoid writing the same processing logic on both Hadoop and Samza, we created
+a library called Starfire to help generate workflows on both batch and stream
```

Comment:
Is Starfire open-source or proprietary? I tried to search for it and dint find. May be make it explicit that its not available openly [yet]